doc/development/architecture.rst

```diff
@@ -4,4 +4,113 @@
 Architecture
 ============
 
-.. image:: /_static/Spicy-Architecture.svg
+Components & Data Flow
+----------------------
+
+.. image:: /_static/architecture.svg
+
+Runtime Libraries
+-----------------
+
+HILTI and Spicy each come with their own runtime libraries providing
+functionality that the execution of compiled code requires. The bulk
+of the functionality here resides with the HILTI side, with the Spicy
+runtime adding pieces that are specific to its use case (i.e.,
+parsing).
+
+Conceptually, there are a few different categories of functionality
+going into these runtime libraries, per the following summary.
+
+.. rubric:: Categories of Functionality
+
+.. image:: /_static/runtime-libraries.svg
+
+Category 1
+    Public library functionality that Spicy programs can ``import``
+    (e.g., functions like ``spicy::current_time()`` inside the
+    ``spicy`` module; filters like ``filter::Zlib`` inside the
+    ``filter`` module). This functionality is declared in
+    ``spicy/lib/*.spicy`` and implemented in C++ in ``libspicy-rt.a``.
+
+Category 2
+    Public library functionality that HILTI programs can ``import``
+    (e.g., the ``hilti::print()`` function inside the ``hilti``
+    module). This functionality is declared in ``hilti/lib/hilti.hlt``
+    and implemented in C++ in ``libhilti-rt.a``.
+
+    .. note::
+
+        "Public functionality" here means being available to any
+        *HILTI* program. This functionality is *not* exposed inside
+        Spicy, and hence usually not visible to users unless they
+        happen to start writing HILTI programs (e.g., when adding test
+        cases to the code base).
+
+Category 3
+    Private Spicy-side library functionality that the HILTI code
+    coming out of Spicy compilation can ``import`` (e.g., functions to
+    access parsing input, such as ``spicy_rt::waitForInput()``;
+    HILTI-side type definitions for Spicy-specific types, such as for
+    a ``sink``). This functionality is declared in
+    ``spicy/lib/spicy_rt.spicy`` and implemented in C++ in
+    ``libspicy-rt.a``.
+
+Category 4
+    Public library functionality for C++ host applications to
```

Comment:
Can we change to ordering of these? With that 1-3 would all be public functionality and 4-5 private.

Reply:
Ok, I was debating about that, will do.

Reply:
Done